### ISSUE TITLE
Deprecation warnings from ActiveRecord::AttributeMethods::Dirty when calling with_lock

### ISSUE THREAD

**Issue by heaven** [rails/rails#30307]
I have 2 models: `Account` and `PayoutMethod` where an account has many payout methods.

In `PayoutMethod` I have the following code :
```ruby
around_save :ensure_primary

private

def ensure_primary
  self.account.with_lock do
    # Make sure we always have a primary method
    # ...

    # Save
    yield
  end
end
```

Every time when I create account together with a payment method the code above produces the following warnings:
```
DEPRECATION WARNING: The behavior of `changed?` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_changes?` instead. (called from set_primary at app/models/market/payout_method.rb:80)
DEPRECATION WARNING: The behavior of `changed_attributes` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_changes.transform_values(&:first)` instead. (called from set_primary at app/models/market/payout_method.rb:80)
DEPRECATION WARNING: Locking a record with unpersisted changes is deprecated and will raise an exception in Rails 5.2. Use `save` to persist the changes, or `reload` to discard them explicitly. (called from set_primary at app/models/market/payout_method.rb:80)
```

The warning points to line 80, it is `self.account.with_lock do`.

Rails 5.1.3

**Comment by gsamokovarov** [comment]
I think that's the correct line that triggers the deprecation warnings.

**Comment by heaven** [comment]
It seems `#lock!` checks for `#changed?` status while we're still in a transaction but `#save` has already been called on `self.account` object.

**Comment by heaven** [comment]
Another example that trigger this warning:
```ruby
class TaggerObserver < ActiveRecord::Observer
  observe []

  def after_commit(record)
    return if record.destroyed?

    record.with_lock do
      # ...
    end
  end
end
```
Tells me
```ruby
DEPRECATION WARNING: Locking a record with unpersisted changes is deprecated and will raise an exception in Rails 5.2. Use `save` to persist the changes, or `reload` to discard them explicitly. (called from after_commit at app/models/tagger_observer.rb:7)
```

And why at all is it illegal to lock record with unpersisted changes? As far as I know, `#with_lock` does reload the record adding `FOR UPDATE` to the query.

Why should I reload it manually and then reload it a second time when using `with_lock`?
```
2.3.3 :020 > profile.reload.with_lock {}
  Profile Load (0.5ms)  SELECT  `profiles`.* FROM `profiles` WHERE `profiles`.`id` = 2485214 LIMIT 1
   (0.2ms)  BEGIN
  Profile Load (0.4ms)  SELECT  `profiles`.* FROM `profiles` WHERE `profiles`.`id` = 2485214 LIMIT 1 FOR UPDATE
   (0.1ms)  COMMIT
```

After upgrade to Rails 5, it feels pretty much like it is trying to do things that not in its competency.

--
If someone doesn't know how to use Rails and does not save record before calling `#with_lock`, should this ever be considered as a reason for such restrictions? Is it Rails responsibility at all? No need to read docs, let's adjust the whole word for noobs so they never suffer. Sorry for this, but I really didn't understand this change.

**Comment by JasonBarnabe** [comment]
Testcase showing the deprecation warning when using `with_lock` in an `after_save` callback:

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", '5.1.4'
  gem "sqlite3"
end

require "active_record"
require "active_support"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)
ActiveSupport::Deprecation.behavior = :raise

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.string :title
  end
end

class Post < ActiveRecord::Base
  after_save do
    with_lock do
      # Stuff
    end
  end
end

class BugTest < Minitest::Test
  def test_callback_with_lock
    Post.create!(title: 'foo')
  end
end
```

> Error:
> BugTest#test_callback_with_lock:
> ActiveSupport::DeprecationException: DEPRECATION WARNING: The behavior of `changed?` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_changes?` instead. (called from block in <class:Post> at active_record_master.rb:37)

**Comment by rafaelfranca** [comment]
That deprecation is by design. the lock there will discard all the changes made before the callback so you should not be calling that inside a callback.

> And why at all is it illegal to lock record with unpersisted changes? As far as I know, #with_lock does reload the record adding FOR UPDATE to the query.

Because that behavior is surprising to some people, so instead of doing this automatically we are asking you to do it. In Rails 5.2 this will raise an exception.

**Comment by heaven** [comment]
@rafaelfranca I think you did not check the code well, the warning appeared when using with_lock inside an `after_commit` callback. I don't see why should I use `reload` there, so this behavior looks completely wrong. In Rails 5.2 a lot os valid code will stop working because of this.

**Comment by rafaelfranca** [comment]
It sounds like `lock` should check `has_changes_to_save?` instead of `changed?` since if it is in a `after_commit` it is already outside the transaction.

@sgrif is that correct?

**Comment by CJStadler** [comment]
It appears that the deprecation warnings from `ActiveRecord::AttributeMethods::Dirty` have already been removed from master in https://github.com/rails/rails/commit/020abadf047997cb3df18a59d210dfe4406cf166. I believe this means these deprecations should no longer be an issue on 5.2. And `changed?` and `has_changes_to_save?` will then have the same behavior, so no changes to `lock!` should be necessary?

I am attempting to update an app to 5.1 though and there doesn't seem to be a solution except waiting for 5.2. Is it possible to update 5.1 to use `has_changes_to_save?`? To do this should I open a PR against the `5.1-stable` branch?

**Comment by kamipo** [comment]
Fixed by #30956.
